implementation.md

```diff
@@ -39,4 +39,25 @@ After examining Compound and Aave Solidity code, I have some thoughts about how
 - You can see a similar concept on Etherscan transactions. It will show you the USD value the moment of the transaction, but it will also provide you with the current USD value if you want.
 
 ## Counts of Events
-- One day `graph-node` will have better aggregation queries, but for now the best way to get a count of events is to implement counters, which I have done for `Protocol`, `Market`, `Account`.
\ No newline at end of file
+- One day `graph-node` will have better aggregation queries, but for now the best way to get a count of events is to implement counters, which I have done for `Protocol`, `Market`, `Account`.
+
+## Open Questions
+- Credit delegation can be implemented by separating the `sender` of transactions apart from the `from` or `to`, and also tracking the function calls that create delegation. But, every protocol implements this differently, and some not at all.
+  - The question is, how to implement this properly? It involves understanding how it works across all protocols and making sure that no events are screwed, as this can confuse the subgraph.
+- Should gasPrice or transactionFee be included?
+- How do Aave stable and variable debt tokens fit in here? Need to examine the code more.
+- Should each protocol extend each event, such as `CompoundDeposit`, `AaveBorrow`, etc? It seems it might be worth it, but might add a lot of indexing time.
+- How much historical data should be included? For example, a user's historical balance for borrows and deposits of their whole portfolio. This, of course would take a ton of indexing, such as how the uniswap.info subgraph takes a long time to sync. 
+- Is there any existing Subgraph standard out there?
+- Can we just filter queries for events like `Deposit` on entity types, and remove most usage of `@derivedFrom`?
+  - In some places yes, like Market and Protocol because the query is straight forward.
+  - But in Account , it is harder because of `to` and `from` both being used.
+  - For now, we are leaving them in.
+
+## Future Work
```

Comment:
The current schema is designed very much as a point-in-time view of the protocol. We should probably consider adding time-series for certain fields/entities so that it'd be easy to retrieve historical data. The Uniswap's subgraph has a set of convenience entities called snapshots that allows one to easily retrieve time-series data without having to do time-travel queries at each individual timestamp.

Reply:
Yes, we should! A few things to consider:

- I think this might be a good situation where we would create a different subgraph, _Lending Core_ and _Lending History_ or something like that.
- I think we should wait on developing any of the mappings for this part of the code. A few reasons:
  - I would like to see the indexing speed without the historical calculations. Historical calculations increase a lot of time for indexing speed.
  - There is an Aggregation Feature being worked on at Edge and Node, see https://hackmd.io/d1XUlIOUSqulUFbtQbs0iw. It might be a few months away still, but we should consider it in our development process.
  
 I think we could go ahead and think through the entities needed for historical data for now, and then sometime in the future, we can implement them when we have a better idea of the indexing speed and the Aggregation Feature. What do you think?

Reply:
If you need this data, we can also work on it now if desired. It will exist eventually, either in the base subgraph, the _Lending History_ subgraph, or a periphery subgraph. We can just put it in the base for now

Reply:
Wow this is cool! I did not know of this. I will definitely need this data but I'm fine leaving it unimplemented for now, see what the indexing time looks like, and come back to it.